tests/cases/compiler/classPropInitializationInferenceWithElementAccess.ts

```diff
@@ -0,0 +1,19 @@
+// @noEmit: true
+// @strict: true
+
+// repro #49339
+export class Cls {
+    x;
+    y;
+    z;
+    
+    0;
+
+    constructor(seed: number) {
+        this['x'] = [seed];
+        this['y'] = { seed };
+        this['z'] = `${seed}`;
```

Comment:
Any kind of non-identifier expression was creating a problem here - probably because of the contextual typing (as mentioned in the issue).